### PR TITLE
Powerhead page show all

### PR DESCRIPTION
**Pull request by petuomin**
* Combines all tribes into single report for "show all"
* Makes people boxes smaller, if there are more than 52 to show.

New version will work to about 300 people boxes. After that we might need to make the boxes even smaller.

### CODE REVIEW

File: src/components/pages/PowerheadPage/model.js [hunk 1]
```diff
@@ -67,7 +67,15 @@ function makeFilterByLocationFn$(selectedLocation$) {
...
-      return {...oldState, reports: newReports};
+
+      return {
+        ...oldState,
+        reports: newReports,
+        filters: {
+          ...oldState.filters,
+          location: location,
+        },
+      };
```
Comment (staltz) [inline src/components/pages/PowerheadPage/model.js]: What are you solving here? I can't really see

File: src/components/pages/PowerheadPage/reportUtils.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+function augmentReportsWithMetadata(reports) {
+  const reportsWithRespectiveTotals = augmentReportsWithFinancials(reports);
+  const maxFinancialsValPerReport = reportsWithRespectiveTotals
+    .map(report => report.maxFinancialsVal);
+  const companyWideMaxFinancialsVal = Math.max(...maxFinancialsValPerReport);
+  return reportsWithRespectiveTotals.map(report => {
+    return {...report, companyWideMaxFinancialsVal};
+  });
+}
```
Comment (staltz) [inline src/components/pages/PowerheadPage/reportUtils.js]: :+1: for this refactor

File: src/components/pages/PowerheadPage/test.js [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+import select from 'vtree-select';
+import {combineReports, augmentReportsWithMetadata} from './reportUtils';
+import _ from 'lodash';
+
+const mockData = {"count":2,"next":null,"previous":null,"results":[{"id":1,"months":[{"September":"2015-09-01"},{"October":"2015-10-01"},{"November":"2015-11-01"},{"December":"2015-12-01"}],"value_creation":[1000,1000,1000,1000],"orderbook":[10000,10000,10000,10000],"overrun":[5,5,5,5],"business_days":[22,22,21,21],"fte":[10,10,10,10],"bench":[2,2,2,2],"ext_fte":[1.0,1.0,1.0,1.0],"name":"Test-org-1","expenses_per_fte_month":"5000.00","days_per_week":5,"hours_per_day":"7.50","target_hourly_rate":"10.00","country":"FI","holiday_calendar":1,"site":1},{"id":5,"months":[{"September":"2015-09-01"},{"October":"2015-10-01"},{"November":"2015-11-01"},{"December":"2015-12-01"}],"value_creation":[2000,2000,2000,2000],"orderbook":[20000,20000,20000,20000],"overrun":[10,10,10,10],"business_days":[22,22,21,21],"fte":[20,20,20,20],"bench":[3,3,3,3],"ext_fte":[2.0,2.0,2.0,2.0],"name":"Test-org-2","expenses_per_fte_month":"4000.00","days_per_week":5,"hours_per_day":"8.00","target_hourly_rate":"10.00","country":"DE","holiday_calendar":2,"site":3}]};
```
Comment (staltz) [inline src/components/pages/PowerheadPage/test.js]: Can we put this in a separate file and formatted nicely as JSON?

File: src/components/pages/PowerheadPage/test.js [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+		expect(augmentedReports[0].maxFinancialsVal).to.eql(55000);
+		expect(augmentedReports[1].maxFinancialsVal).to.eql(88000);
+	});
+
+	it('should add companyWideMaxFinancialsVal to reports', () => {
+		const augmentedReports = augmentReportsWithMetadata(mockData.results);
+		expect(augmentedReports[0].companyWideMaxFinancialsVal).to.eql(88000);
+		expect(augmentedReports[1].companyWideMaxFinancialsVal).to.eql(88000);
+	});
+});
```
Comment (staltz) [inline src/components/pages/PowerheadPage/test.js]: :+1: beautiful tests

File: src/components/pages/PowerheadPage/test.js [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+	const props$ = Rx.Observable.just({});
+	const HTTPSource = Rx.Observable.empty();
+	const {request$, response$} = PowerheadPageHTTP({props$, HTTP: HTTPSource});
+	
+	request$.subscribe(req => {
+
+		expect( _.startsWith(req.url, `${API_PATH}/powerhead/`) ).to.be.true;
+		done();
+	});
+  });
```
Comment (staltz) [inline src/components/pages/PowerheadPage/test.js]: Bit crazy code style going on here. Did the ESLint not complain?

File: src/components/pages/PowerheadPage/test.js [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+	request$.map(req => {
+		
+		var response$ = Rx.Observable.just({
+			request: req,
+			body: { results: "test" }
+		});
+		response$.request = req;
+		return response$;
+
+	}).subscribe(HTTPSource.asObserver());
```
Comment (staltz) [inline src/components/pages/PowerheadPage/test.js]: Correct approach to mock HTTPSource. :+1:

File: src/components/pages/PowerheadPage/test.js [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+ 		it('should have a sum overruns from all reports, for each month', () => {
+ 			expect(page.overruns).to.eql(['15','15','15']);
+ 		});
+
+    });
+
+
+  });
+});
```
Comment (staltz) [inline src/components/pages/PowerheadPage/test.js]: Overall, great tests. Thanks.
We just need to fix some ESLint issues and that's it.

File: src/components/pages/PowerheadPage/view.js [hunk 8]
```diff
@@ -40,20 +42,21 @@ function renderPeopleStatsList(report, monthIndex) {
...
-      {renderPeopleStatsItem('Bench', benchVal, 'FTE')}
+      {renderPeopleStatsItem('Total exts.', extFteVal, 'FTE', 'exts', true)}
+      {renderPeopleStatsItem('Total ints.', totalIntsVal, 'FTE', 'ints')}
+      {renderPeopleStatsItem('Booked', bookedVal, 'FTE', 'booked')}
+      {renderPeopleStatsItem('Bench', benchVal, 'FTE', 'bench')}
     </ul>
   );
 }
 
-function renderFinancialsStatsItem(label, value, legendBarStyle, special = false) {
```
Comment (staltz) [inline src/components/pages/PowerheadPage/view.js]: 5 arguments starts to smell bad, but I'm not really sure what to do.
At least we should try to combine legendBarStyle and special into just one `style` argument which can be set on the `<li>`, not on the div for the legend. So then in SCSS we can set both the legend style and the span styles by using nested SCSS rules.

Comment (staltz) [inline src/components/pages/PowerheadPage/view.js]: But you don't need to update this PR. Let's merge this and do the refactor separately in another PR.

File: src/components/pages/PowerheadPage/view.js [hunk 9]
```diff
@@ -69,32 +72,56 @@ function renderFinancialsStatsList(report, monthIndex) {
...
-      return <li className={styles.monthGraphPeopleBoxBooked}></li>;
+      return <li className={bookedBoxStyle}></li>;
     } else {
-      return <li className={styles.monthGraphPeopleBoxBench}></li>;
+      return <li className={benchBoxStyle}></li>;
     }
   });
-  const chunks = _.chunk(boxes, 20);
+
+  const chunks = _.chunk(boxes, (totalPeople > useSmallBoxesThreshold) ? 40 : 20);
```
Comment (staltz) [inline src/components/pages/PowerheadPage/view.js]: Instead of 40 and 20 we should name those as constants.

File: src/components/pages/PowerheadPage/view.js [hunk 10]
```diff
@@ -69,32 +72,56 @@ function renderFinancialsStatsList(report, monthIndex) {
...
+      )}
+      {renderFinancialsStatsItem(
+        'Confirmed revenue', revenue, revenueBarStyle, 'confirmed_revenue'
+      )}
     </ul>
   );
 }
 
 function renderMonthGraphPeople(report, monthIndex) {
+  const useSmallBoxesThreshold = 52;
```
Comment (staltz) [inline src/components/pages/PowerheadPage/view.js]: Not for this PR, but I think a better approach would be to dynamically stretch in and out, instead of a magic-number threshold approach. E.g. If a tribe suddenly has 53 people (quite possible), it would suddenly look not so nice.

File: src/components/pages/PowerheadPage/view.js [hunk 11]
```diff
@@ -238,21 +241,30 @@ function renderLoadingIndicator() {
...
+        reports = combineReports(state.reports);
+      } else {
+        const sortedReports = sortReportsByLargestTribe(state.reports);
+        const completeReports = augmentReportsWithMetadata(sortedReports);
+        reports = completeReports;
       }
-    </div>
-  );
+
+      return <div>
```
Comment (staltz) [inline src/components/pages/PowerheadPage/view.js]: usually we do `return (` `newline` `<div>` as the code style.

File: src/components/pages/PowerheadPage/model.js (hunk 1 above)
Comment (petuomin) [inline src/components/pages/PowerheadPage/model.js]: The filter updates the reports in the state, but before this PR it would not update the filter values. I'm checking in the [view](https://github.com/futurice/power-ui/blob/powerhead-page-show-all/src/components/pages/PowerheadPage/view.js#L246) wheter `location === 'all'` and create a combined report if true. So the view needs not only the filtered reports but also what filters were used to create the filtered list. I think it would make sense for other filters to add this data too.

File: src/components/pages/PowerheadPage/test.js (hunk 5 above)
Comment (petuomin) [inline src/components/pages/PowerheadPage/test.js]: We have .eslintignore file that matches all of our test files. Maybe we should make a branch where we would remove the ignore and fix style errors in tests. We can always override certain rules for tests.
